### DISCUSSION TITLE
Easy contributions! Add typehint and translate docs for MMDetection3d~

### DISCUSSION THREAD

**Discussion by JingweiZhang12** [open-mmlab/mmdetection3d#1982] [General]
Hi guys, we are planning to **add typehint** for python code and **translate docs** from English to Chinese in MMDetection3d. PRs from the community are welcomed :stuck_out_tongue_closed_eyes: . To make the community fully aware of the progress, we list the progress here. Please feel free to leave a message and create a PR if you are willing to do these :rocket::rocket:.

## Typehint (better to check the docstring)
- [x] datasets @Huangtianyu-caas 
    + [ ] [data_augment_utils.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/transforms/data_augment_utils.py)
   + [x] [compose.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/transforms/compose.py)
   + [x] [dataset_wrappers.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/dataset_wrappers.py)
   + [x] [kitti2d_dataset.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/kitti2d_dataset.py)
   + [x] [s3dis_dataset.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/s3dis_dataset.py) @Xiangxu-0103 https://github.com/open-mmlab/mmdetection3d/pull/1984
- [ ] evaluation @gaopeng91 
   + [ ] [eval.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/kitti_utils/eval.py)
   + [ ] [rotate_iou.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/kitti_utils/rotate_iou.py)
   + [ ] [evaluate_semantic_instance.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/scannet_utils/evaluate_semantic_instance.py)
   + [ ] [util_3d.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/scannet_utils/util_3d.py)
   + [ ] [prediction_kitti_to_waymo.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/waymo_utils/prediction_kitti_to_waymo.py)
   + [ ] [indoor_eval.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/indoor_eval.py)
   + [ ] [instance_seg_eval.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/instance_seg_eval.py)
   + [ ] [lyft_eval.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/lyft_eval.py)
   + [ ] [seg_eval.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/evaluation/functional/seg_eval.py)
- [ ] structure part-1
   + [ ] [structure/bbox_3d/](https://github.com/open-mmlab/mmdetection3d/tree/dev-1.x/mmdet3d/structures/bbox_3d)
- [ ] structure part-2
   + [ ] [ops/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/structures/ops)
   + [ ] [points/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/structures/points)
- [x] models part-1
   + [x] [layers/dgcnn_modules/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/dgcnn_modules) @Xiangxu-0103 https://github.com/open-mmlab/mmdetection3d/pull/2014
   + [x] [layers/fusion_layers/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/fusion_layers) @Xiangxu-0103 https://github.com/open-mmlab/mmdetection3d/pull/2014
   + [x] [write_spconv2.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/spconv/overwrite_spconv/write_spconv2.py)
- [x] models part-2 @Xiangxu-0103 https://github.com/open-mmlab/mmdetection3d/pull/2014
   + [x] [layers/paconv/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/paconv)
   + [x] [layers/pointnet_modules/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/pointnet_modules)
- [x] models part-3
   + [x] [box3d_nms.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/box3d_nms.py)
   + [x] [edge_fusion_module.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/edge_fusion_module.py)
   + [x] [mlp.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/mlp.py)
   + [x] [norm.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/norm.py)
   + [x] [sparse_block.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/sparse_block.py)
   + [x] [transformer.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/transformer.py)
   + [x] [vote_module.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/layers/vote_module.py)
- [ ] models part-4
   + [ ] [backbones/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/backbones)
- [x] models part-5
   + [x] [denseheads/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/decode_heads)
   + [x] [groupfree3dnet.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/detectors/groupfree3dnet.py)
   + [x] [dfm.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/detectors/dfm.py)
   + [x] [multiview_dfm.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/detectors/multiview_dfm.py)
   + [x] [ssd3dnet.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/detectors/ssd3dnet.py)
- [x] models part-6 @roger-lcc 
   + [x] [losses/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/losses)
- [ ] models part-7
   + [ ] [middle_encoders/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/middle_encoders)
   + [ ] [voxel_encoders/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/voxel_encoders)
- [ ] models part-8
   + [ ] [task_modules/anchor/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/task_modules/anchor)
   + [ ] [task_modules/assigners/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/task_modules/assigners)
   + [ ] [task_modules/samplers/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/task_modules/samplers)
   + [ ] [task_modules/voxel/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/task_modules/voxel)
- [ ] models part-9
   + [ ] [task_modules/coders/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/task_modules/coders)
- [x] models part-10 and others @shufanwu 
   + [x] [utils/](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/utils)
   + [x] [merge_augs.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/test_time_augs/merge_augs.py)
   + [ ] [array_converter.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/utils/array_converter.py)

## Docs translation
- [ ] [changelog.md](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/docs/en/notes/changelog.md)
- [x] [changelog_v1.0.x.md](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/docs/en/notes/changelog_v1.0.x.md) @song-xyz https://github.com/open-mmlab/mmdetection3d/pull/2075
- [ ] [compatibility.md](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/docs/en/notes/compatibility.md)

**Comment by Xiangxu-0103** [comment]
Hi, from my perspective, [compose.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/transforms/compose.py) can be removed after #1984 being merged, which has supported detection task on `S3DIS dataset`, and [kitti2d_dataset.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/kitti2d_dataset.py) can also be removed for unified using in [kitti_dataset.py](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/datasets/kitti_dataset.py).

**Reply by JingweiZhang12** [reply]
Thanks for your kind remind.

**Comment by Xiangxu-0103** [comment]
For `S3DIS` and other datasets, it has been fixed in #1984.
For `models/layers`, I have added typehint in #2014.

**Comment by moondiana** [comment]
moondiana+models part-4

**Comment by gaopeng91** [comment]
高鹏 type hints evaluation

**Comment by roger-lcc** [comment]
roger-lcc models part-6

**Comment by song-xyz** [comment]
doc trans: changelog_v1.0.x

**Comment by shufanwu** [comment]
wushufan models part-10 and others

**Comment by Huangtianyu-caas** [comment]
黄天宇 mmdet3d-46 type hints

**Reply by JingweiZhang12** [reply]
@Huangtianyu-caas  Please take a specific task. For example， part-8

**Reply by Huangtianyu-caas** [reply]
Thank you for your explanation. I would choose a dataset task
